### DISCUSSION TITLE
Can I use firestore browser SDK using firebase auth session cookie?

### DISCUSSION THREAD

**Discussion by nkpgardose** [firebase/firebase-js-sdk#7094] [Q&A]
I'm using sessions/cookies to authenticate my users.

```ts
setPersistence(auth, inMemoryPersistence)
...
const onSignIn = (provider: GoogleAuthProvider) => {
  setLoading(true)
  signInWithPopup(auth, provider)
    .then((userAuth) => {
      /**
       * Generate auth cookie for server page component.
       */
      userAuth.user.getIdToken().then((jwtToken) => {
        fetch('/api/get-auth-cookie', {
          headers: { Authorization: `Bearer ${jwtToken}` },
        })
          .then(() =>
            /**
             * Sign out client auth as the persistence is set to NONE.
             */
            signOut(auth)
          )
          .then(() => {
            flushSync(() => {
              setLoading(false)
            })
            router.push('/home')
          })
          .catch((error) => console.error(error))
      })
    })
    .catch((error) => {
      console.error(error)
    })
}
```

and I'm using `verifySessionCookie()` on the server for some API authentication, perform some tasks, etc.

However, I do want to have the option to run some firestore in the client-side like so

```ts
const onCreateWallet = () => {
  // Add a new document with a generated id.
  addDoc(collection(firestoreDB, 'wallets'), {
    name: 'My daily debit card',
    uid: getAuth().currentUser?.uid,
  })
    .then((docRef) => {
      console.log('Wallet created')
      console.log('Document written with ID: ', docRef.id)
    })
    .catch((error) => {
      console.error('Error adding document: ', error)
    })
}
```
In security rules:
```
function hasValidName() {
  return request.resource.data.name is string &&
    request.resource.data.name.size() > 3 &&
    request.resource.data.name.size() < 30
}

function isValidWallet() {
  return hasValidName() &&
    request.resource.data.uid == request.auth.uid;
}
```

But as far as the `getAuth` and the security rules in place. `currentUser` here is already null.

Would there be a good approach here to use firestore in the client side? and use session cookie to hydrate `currentUser` values and `request.auth.uid` in security rules?

I'd like to have this kind of option as NextJS can be SSR & interactive in client at the same time. I'd appreciate the help.

**Comment by TrinaxCode** [comment]
Because you call `signOut(auth)` after creating the cookie, `getAuth().currentUser` being `null` is expected. The session-cookie flow is for server-side authentication: your server verifies the cookie with the Admin SDK. It does not sign the browser's Firebase Auth client back in, and the Firestore web SDK cannot use that cookie directly for a Firestore request. Firestore Rules populate `request.auth` from Firebase Authentication on the client request. See [Manage session cookies](https://firebase.google.com/docs/auth/admin/manage-cookies) and [Firestore rule conditions](https://firebase.google.com/docs/firestore/security/rules-conditions).

You have two straightforward choices:

1. **Write from the browser under Firestore Rules:** keep the Firebase client signed in (your `inMemoryPersistence` is suitable if the auth state only needs to last for the current page session) and do not call `signOut` after setting the cookie. The Firestore request will then have the signed-in user's UID in `request.auth`.
2. **Keep only the HttpOnly session cookie:** send the wallet write to your own API, verify the cookie there with `verifySessionCookie()`, then write with the Admin SDK. Server Firestore libraries bypass Firestore Rules, so the API must check the verified UID and enforce ownership itself. See [Firestore server client libraries](https://firebase.google.com/docs/firestore/security/rules-query).

If you need a Firebase-authenticated browser client after a cookie-based login, a server can verify the cookie and mint a custom token for that UID; the client can then call `signInWithCustomToken()`. That creates a normal Firebase client auth session, so it is an alternative identity flow rather than using the session cookie directly. See [custom authentication](https://firebase.google.com/docs/auth/web/custom-auth).
